Lazy tensor: automatically promote constants to inputs based on history

bgogul opened the pull request:
This PR implements the logic to promote constants to inputs (https://bugs.swift.org/browse/TF-704). Suppose that we have the following traces:

```
lazyTrace_sum2(%0: float) -> (float) {
   %1 = Const[dtype: float, value: 2.0]() 
   %2 = Add[T: float](%0, %1)
   return %2
}
```

```
lazyTrace_sum2(%0: float) -> (float) {
   %1 = Const[dtype: float, value: 3.0]() 
   %2 = Add[T: float](%0, %1)
   return %2
}
```
The above two functions are different and will need to be compiled individually. Suppose that we promote these constants to inputs as follows:
```
lazyTrace_sum3(%0: float, %1: float) -> (float) {
   %2 = Add[T: float](%0, %1)
   return %2
}
```
Only the inputs change and the function remains constant, which eliminates the additional compilation overhead.

eaplatanios commented on Tests/TensorFlowTests/LazyTensorTraceCacheTests.swift:
```diff
@@ -0,0 +1,131 @@
+@testable import TensorFlow
+import CTensorFlow
+
+final class LazyTensorTraceCacheTests : LazyTensorTestCase {
```
```suggestion
final class LazyTensorTraceCacheTests: LazyTensorTestCase {
```

eaplatanios commented on Tests/TensorFlowTests/LazyTensorTraceCacheTests.swift:
```diff
@@ -0,0 +1,131 @@
+        XCTAssertEqual(y.scalars, [5.0, 10.0, 15.0])
+    }
+
+    func testDontPromoteEqualConstants() {
```
```suggestion
    func testDoNotPromoteEqualConstants() {
```

eaplatanios commented on Tests/TensorFlowTests/LazyTensorTraceCacheTests.swift:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    static var allTests = [
+        ("testAutoConstPromotion", testAutoConstPromotion),
```
```suggestion
        ("testConstantPromotion", testConstantPromotion),
```

eaplatanios commented on Tests/TensorFlowTests/LazyTensorTraceCacheTests.swift:
```diff
@@ -0,0 +1,131 @@
+
+    static var allTests = [
+        ("testAutoConstPromotion", testAutoConstPromotion),
+        ("testDontPromoteEqualConstants", testDontPromoteEqualConstants)
```
```suggestion
        ("testDoNotPromoteEqualConstants", testDoNotPromoteEqualConstants)
```

bgogul commented:
> > @rxwei this is not super urgent, but can you take a look at this when you get a chance?
> 
> Sorry for the delay!

No worries. I addressed your comments and left some clarification questions.
